### ISSUE TITLE
Lint for uppercase typedef and function parameter names

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#921]
What does the following define?

```dart
typedef String camelize(String);

void search(String haystack, callback(String)) {...}
```

When only one identifier is found, it is the _name_ of a parameter, not the _type_. The above is equivalent to:

```dart
typedef String camelize(dynamic String);

void search(String haystack, callback(dynamic String)) {...}
```

How about a lint that detects single-identifier parameters which start with a capital letter?

**Comment by bwilkerson** [comment]
I think that would be great. It might even be better (in the sense of fewer false positives) if the lint looked for identifier names that match a type that is visible in scope.

**Comment by srawlins** [comment]
Oh that's a good restriction. Definitely fewer false positives.

**Comment by munificent** [comment]
Note that the new function type syntax doesn't have this problem:

```dart
typedef camelize = String Function(String);

void search(String haystack, Function(String) callback) {...}
```

**Comment by gabrc** [comment]
Also need to detect single-identifier parameters that equal 'bool', 'int', 'double' and 'num', and possibly 'dynamic'.

**Comment by bwilkerson** [comment]
The `dynamic` case is interesting, but the other types are all normal types that can be handled in the same way as any user-defined types.
